File: php/src/Transport.php
<?php

declare(strict_types=1);

namespace Idkollen\Client;

use Psr\Http\Client\ClientExceptionInterface;
use Psr\Http\Client\ClientInterface;
use Psr\Http\Message\RequestFactoryInterface;
use Psr\Http\Message\RequestInterface;
use Psr\Http\Message\StreamFactoryInterface;

/**
 * @internal Used by endpoint classes. Not part of the public API.
 */
final class Transport
{
    private const USER_AGENT = 'idkollen-client-php/0.1.0';

    public function __construct(
        private readonly ClientInterface $httpClient,
        private readonly RequestFactoryInterface $requestFactory,
        private readonly StreamFactoryInterface $streamFactory,
        private readonly string $baseUrl,
        private readonly string $clientId,
        private readonly string $clientSecret,
    ) {}

    public function post(string $path, array $body): array
    {
        $request = $this->buildRequest('POST', $path)
            ->withHeader('Content-Type', 'application/json')
            ->withBody($this->streamFactory->createStream(json_encode($body, JSON_THROW_ON_ERROR)));

        return $this->send($request);
    }

    public function get(string $path): array
    {
        return $this->send($this->buildRequest('GET', $path));
    }

    public function getRaw(string $path): string
    {
        return $this->sendRaw($this->buildRequest('GET', $path));
    }

    public function delete(string $path): void
    {
        $this->send($this->buildRequest('DELETE', $path));
    }

    public function postMultipart(string $path, string $data, string $filename, string $mimeType): array
    {
        $boundary = bin2hex(random_bytes(16));
        $body = "--{$boundary}\r\n"
            . "Content-Disposition: form-data; name=\"file\"; filename=\"{$filename}\"\r\n"
            . "Content-Type: {$mimeType}\r\n"
            . "\r\n"
            . $data
            . "\r\n"
            . "--{$boundary}--\r\n";

        $request = $this->buildRequest('POST', $path)
            ->withHeader('Content-Type', "multipart/form-data; boundary={$boundary}")
            ->withBody($this->streamFactory->createStream($body));

        return $this->send($request);
    }

    private function buildRequest(string $method, string $path): RequestInterface
    {
        return $this->requestFactory->createRequest($method, $this->baseUrl . $path)
            ->withHeader('User-Agent', self::USER_AGENT)
            ->withHeader('Authorization', 'Basic ' . base64_encode("{$this->clientId}:{$this->clientSecret}"));
    }

    private function send(RequestInterface $request): array
    {
        $body = $this->sendRaw($request);
        if ($body === '') {
            return [];
        }
        $decoded = json_decode($body, true);
        return is_array($decoded) ? $decoded : [];
    }

    private function sendRaw(RequestInterface $request): string
    {
        try {
            $response = $this->httpClient->sendRequest($request);
        } catch (ClientExceptionInterface $e) {
            throw new IdkollenException(0, $e->getMessage());
        }

        $statusCode = $response->getStatusCode();
        $body = (string) $response->getBody();

        if ($statusCode < 200 || $statusCode >= 300) {
            $decoded = json_decode($body, true);
            $message = is_array($decoded) && isset($decoded['message']) ? (string) $decoded['message'] : $body;
            throw new IdkollenException($statusCode, $message);
        }

        return $body;
    }
}

File: dart/lib/src/transport.dart
import 'dart:convert';
import 'package:http/http.dart' as http;
import 'package:http_parser/http_parser.dart';

import 'exceptions.dart';

class Transport {
  Transport({
    required http.Client httpClient,
    required String baseUrl,
    required String clientId,
    required String clientSecret,
  })  : _http = httpClient,
        _baseUrl = baseUrl.endsWith('/') ? baseUrl.substring(0, baseUrl.length - 1) : baseUrl,
        _authHeader = 'Basic ${base64Encode(utf8.encode('$clientId:$clientSecret'))}';

  static const _userAgent = 'idkollen-client-dart/0.1.0';

  final http.Client _http;
  final String _baseUrl;
  final String _authHeader;

  void close() => _http.close();

  Future<Map<String, dynamic>> post(String path, Map<String, dynamic> body) async {
    final res = await _send(http.Request('POST', Uri.parse('$_baseUrl$path'))
      ..headers['Content-Type'] = 'application/json'
      ..body = jsonEncode(body));
    return _decodeJson(res);
  }

  Future<Map<String, dynamic>> get(String path) async {
    final res = await _send(http.Request('GET', Uri.parse('$_baseUrl$path')));
    return _decodeJson(res);
  }

  Future<List<int>> getRaw(String path) async {
    final res = await _send(http.Request('GET', Uri.parse('$_baseUrl$path')));
    return res.bodyBytes;
  }

  Future<void> delete(String path) async {
    await _send(http.Request('DELETE', Uri.parse('$_baseUrl$path')));
  }

  Future<Map<String, dynamic>> postMultipart(
      String path, List<int> data, String filename, String mimeType) async {
    final req = http.MultipartRequest('POST', Uri.parse('$_baseUrl$path'));
    req.headers['Authorization'] = _authHeader;
    req.headers['User-Agent'] = _userAgent;
    req.files.add(http.MultipartFile.fromBytes(
      'file',
      data,
      filename: filename,
      contentType: MediaType.parse(mimeType),
    ));
    final streamed = await _http.send(req);
    final res = await http.Response.fromStream(streamed);
    _ensureOk(res);
    return _decodeJson(res);
  }

  Future<http.Response> _send(http.Request req) async {
    req.headers['Authorization'] = _authHeader;
    req.headers['User-Agent'] = _userAgent;
    http.Response res;
    try {
      final streamed = await _http.send(req);
      res = await http.Response.fromStream(streamed);
    } on http.ClientException catch (e) {
      throw IdkollenException(0, e.message);
    }
    _ensureOk(res);
    return res;
  }

  void _ensureOk(http.Response res) {
    if (res.statusCode < 200 || res.statusCode >= 300) {
      String message;
      try {
        final decoded = jsonDecode(res.body);
        message = (decoded is Map && decoded['message'] is String) ? decoded['message'] as String : res.body;
      } catch (_) {
        message = res.body;
      }
      throw IdkollenException(res.statusCode, message);
    }
  }

  Map<String, dynamic> _decodeJson(http.Response res) {
    if (res.body.isEmpty) return <String, dynamic>{};
    final decoded = jsonDecode(res.body);
    return decoded is Map<String, dynamic> ? decoded : <String, dynamic>{};
  }
}
